### PR TITLE
New Resource: aws_appautoscaling_scheduled_action

### PR DESCRIPTION
**Pull request by atsushi-ishibashi**
#2227

### CODE REVIEW

**Comment by atsushi-ishibashi** [comment]
```
make testacc TEST=./aws TESTARGS='-run=TestAccAwsAppautoscalingScheduledAction_dynamo'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAwsAppautoscalingScheduledAction_basic -timeout 120m
=== RUN   TestAccAwsAppautoscalingScheduledAction_basic
--- PASS: TestAccAwsAppautoscalingScheduledAction_basic (90.87s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	90.907s 
```
```
make testacc TEST=./aws TESTARGS='-run=TestAccAwsAppautoscalingScheduledAction_ECS'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAwsAppautoscalingScheduledAction_ECS -timeout 120m
=== RUN   TestAccAwsAppautoscalingScheduledAction_ECS
--- PASS: TestAccAwsAppautoscalingScheduledAction_ECS (122.79s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	122.848s
```
```
make testacc TEST=./aws TESTARGS='-run=TestAccAwsAppautoscalingScheduledAction_EMR'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAwsAppautoscalingScheduledAction_EMR -timeout 120m
=== RUN   TestAccAwsAppautoscalingScheduledAction_EMR
--- PASS: TestAccAwsAppautoscalingScheduledAction_EMR (701.34s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	701.407s
```
```
make testacc TEST=./aws TESTARGS='-run=TestAccAwsAppautoscalingScheduledAction_SpotFleet'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAwsAppautoscalingScheduledAction_SpotFleet -timeout 120m
=== RUN   TestAccAwsAppautoscalingScheduledAction_SpotFleet
--- PASS: TestAccAwsAppautoscalingScheduledAction_SpotFleet (330.93s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	330.970s
```

File: aws/resource_aws_appautoscaling_scheduled_action.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+			"scalable_dimension": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+					validDimensions := []string{"ecs:service:DesiredCount", "ec2:spot-fleet-request:TargetCapacity",
+						"elasticmapreduce:instancegroup:InstanceCount", "appstream:fleet:DesiredCapacity",
+						"dynamodb:table:ReadCapacityUnits", "dynamodb:table:WriteCapacityUnits",
+						"dynamodb:index:ReadCapacityUnits", "dynamodb:index:WriteCapacityUnits"}
```
Comment (radeksimko) [inline aws/resource_aws_appautoscaling_scheduled_action.go]: While offline, plan-time validation is usually useful I'm afraid it could be counter-productive here as the list of valid dimensions may and will expand and it did recently. I think we'd play a constant catch-up with AWS here and rushing to cut a release just to support a new dimension just because of this strict validation.
Theoretically we could validate for something like `[a-zA-Z]+:[a-zA-Z]+:[a-zA-Z]+` but frankly I'd just rather avoid the validation altogether here.

File: aws/resource_aws_appautoscaling_scheduled_action.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+				Required: true,
+				ForceNew: true,
+			},
+			"service_namespace": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+					validServiceNamespaces := []string{"ecs", "elasticmapreduce", "ec2", "appstream", "dynamodb"}
```
Comment (radeksimko) [inline aws/resource_aws_appautoscaling_scheduled_action.go]: As below - I'm afraid the list of valid service namespaces may/will grow and I'd like us to avoid the maintenance burden associated with it, if possible.

File: aws/resource_aws_appautoscaling_scheduled_action.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+			},
+			"arn": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsAppautoscalingScheduledActionCreate(d *schema.ResourceData, meta interface{}) error {
```
Comment (radeksimko) [inline aws/resource_aws_appautoscaling_scheduled_action.go]: Per [docs](http://docs.aws.amazon.com/cli/latest/reference/application-autoscaling/put-scheduled-action.html) I think we should support updates of this resource and basically just rename this function to something like `resourceAwsAppautoscalingScheduledActionCreateOrUpdate` or `resourceAwsAppautoscalingScheduledActionPut` and reference it in `Update` too.

File: aws/resource_aws_appautoscaling_scheduled_action.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+	err := resource.Retry(5*time.Minute, func() *resource.RetryError {
+		_, err := conn.PutScheduledAction(input)
+		if err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				switch aerr.Code() {
+				case applicationautoscaling.ErrCodeObjectNotFoundException:
+					return resource.RetryableError(err)
+				default:
+					return resource.NonRetryableError(err)
+				}
```
Comment (radeksimko) [inline aws/resource_aws_appautoscaling_scheduled_action.go]: Q: Is this related to eventual consistency?

Regardless of the reason I think the above conditional starting on L153 can be simplified to something like this:

```go
if err != nil {
  if isAWSErr(err, applicationautoscaling.ErrCodeObjectNotFoundException, "") {
    return resource.RetryableError(err)
  }
  return resource.NonRetryableError(err)
}
```

File: aws/resource_aws_appautoscaling_scheduled_action.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+	resp, err := conn.DescribeScheduledActions(input)
+	if err != nil {
+		return err
+	}
+	if len(resp.ScheduledActions) < 1 {
+		d.SetId("")
+		return nil
+	}
+	if len(resp.ScheduledActions) != 1 {
+		return fmt.Errorf("Number of Scheduled Action (%s) isn't one, got %d", saName, len(resp.ScheduledActions))
```
Comment (radeksimko) [inline aws/resource_aws_appautoscaling_scheduled_action.go]: Grammar nitpick, do you mind changing this to `Expected 1 scheduled action under %s, found %d`?

File: aws/resource_aws_appautoscaling_scheduled_action.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	d.SetId(d.Get("name").(string))
```
Comment (radeksimko) [inline aws/resource_aws_appautoscaling_scheduled_action.go]: Shouldn't the ID consist of name + service namespace as that's how we look it up? In other words my question is: Can there be more than 1 scheduled action of the same name? If not then ignore me.

File: aws/resource_aws_appautoscaling_scheduled_action_test.go [hunk 7]
```diff
@@ -0,0 +1,679 @@
...
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ecs:DescribeServices",
+        "ecs:UpdateService",
+				"cloudwatch:DescribeAlarms"
```
Comment (radeksimko) [inline aws/resource_aws_appautoscaling_scheduled_action_test.go]: Nitpick - extra whitespace here 👀

File: aws/resource_aws_appautoscaling_scheduled_action_test.go [hunk 8]
```diff
@@ -0,0 +1,679 @@
...
+        "*"
+      ]
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_ecs_cluster" "hoge" {
+	name = "tf-ecs-cluster-%s"
```
Comment (radeksimko) [inline aws/resource_aws_appautoscaling_scheduled_action_test.go]: extra whitespace here 👀

File: aws/resource_aws_appautoscaling_scheduled_action_test.go [hunk 9]
```diff
@@ -0,0 +1,679 @@
...
+}
+
+resource "aws_ecs_task_definition" "hoge" {
+  family = "foobar"
```
Comment (radeksimko) [inline aws/resource_aws_appautoscaling_scheduled_action_test.go]: Do you mind using a random name here too, please?

File: website/docs/r/appautoscaling_scheduled_action.html.markdown [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the scheduled action.
+* `service_namespace` - (Required) The namespace of the AWS service. Documentation can be found in the parameter at: [AWS Application Auto Scaling API Reference](https://docs.aws.amazon.com/ApplicationAutoScaling/latest/APIReference/API_PutScheduledAction.html#ApplicationAutoScaling-PutScheduledAction-request-ServiceNamespace) Valid Values: ecs | elasticmapreduce | ec2 | appstream | dynamodb
```
Comment (radeksimko) [inline website/docs/r/appautoscaling_scheduled_action.html.markdown]: I would avoid documenting all valid values as that will become a maintenance burden as these expand. How about we just mention a few examples? The user can always IMO look up all valid values in the docs, right?

File: website/docs/r/appautoscaling_scheduled_action.html.markdown [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+* `resource_id` - (Required) The identifier of the resource associated with the scheduled action. Documentation can be found in the parameter at: [AWS Application Auto Scaling API Reference](https://docs.aws.amazon.com/ApplicationAutoScaling/latest/APIReference/API_PutScheduledAction.html#ApplicationAutoScaling-PutScheduledAction-request-ResourceId)
+* `scalable_dimension` - (Optional) The scalable dimension. Documentation can be found in the parameter at: [AWS Application Auto Scaling API Reference](https://docs.aws.amazon.com/ApplicationAutoScaling/latest/APIReference/API_PutScheduledAction.html#ApplicationAutoScaling-PutScheduledAction-request-ScalableDimension) Valid Values: ecs:service:DesiredCount | ec2:spot-fleet-request:TargetCapacity | elasticmapreduce:instancegroup:InstanceCount | appstream:fleet:DesiredCapacity | dynamodb:table:ReadCapacityUnits | dynamodb:table:WriteCapacityUnits | dynamodb:index:ReadCapacityUnits | dynamodb:index:WriteCapacityUnits
```
Comment (radeksimko) [inline website/docs/r/appautoscaling_scheduled_action.html.markdown]: I would avoid documenting all valid values as that will become a maintenance burden as these expand. How about we just mention a few examples? The user can always IMO look up all valid values in the docs, right?

File: aws/resource_aws_appautoscaling_scheduled_action.go (hunk 4 above)
Comment (atsushi-ishibashi) [reply aws/resource_aws_appautoscaling_scheduled_action.go]: I assumed that soon after creating dynamo table and on...

File: aws/resource_aws_appautoscaling_scheduled_action.go (hunk 6 above)
Comment (atsushi-ishibashi) [reply aws/resource_aws_appautoscaling_scheduled_action.go]: According to `DeleteScheduledActionInput` docs, `ResourceId, ScheduledActionName, ServiceNamespace` are required so we have to generate id from these three.
Should I convert to hash?

File: aws/resource_aws_appautoscaling_scheduled_action.go (hunk 2 above)
Comment (atsushi-ishibashi) [reply aws/resource_aws_appautoscaling_scheduled_action.go]: Ok, I removed

File: aws/resource_aws_appautoscaling_scheduled_action.go (hunk 1 above)
Comment (atsushi-ishibashi) [reply aws/resource_aws_appautoscaling_scheduled_action.go]: [docs](https://docs.aws.amazon.com/ApplicationAutoScaling/latest/APIReference/API_PutScheduledAction.html#ApplicationAutoScaling-PutScheduledAction-request-ScalableDimension) lists constant strings so I removed `ValidateFunc`.

File: aws/resource_aws_appautoscaling_scheduled_action.go (hunk 3 above)
Comment (atsushi-ishibashi) [reply aws/resource_aws_appautoscaling_scheduled_action.go]: OK👍

File: aws/resource_aws_appautoscaling_scheduled_action.go (hunk 6 above)
Comment (radeksimko) [reply aws/resource_aws_appautoscaling_scheduled_action.go]: I'd suggest we make it something like this:

```go
d.SetId(name+"-"+serviceNamespace+"-"+resourceId)
```
assuming the underscore has no special meaning in either of these fields.
